network-area-diagram-viewer/src/main/resources/svg.js

```diff
@@ -0,0 +1,463 @@
+
+// TODO(Luma) consider three winding transformers
+// TODO(Luma) consider nodes with multiple buses
+// TODO(Luma) separate components of edges, having one that can be "scaled" (length line from exit of node to middle point)
+//              for reference, see Android NinePatch drawables
+//              https://developer.android.com/develop/ui/views/graphics/drawables#nine-patch
+// TODO(Luma) JavaFX seems to have problems drawing a translated foreign object, try to enclose it in a <g> element?
+// TODO(Luma) update on client: find related hidden nodes and update initial position in metadata?
+
+window.addEventListener('load', function() {
+    var svgDiagram = document.getElementsByTagName("svg")[0];
+    if (svgDiagram != null) {
+        makeDraggableSvg(svgDiagram);
+    }
+});
+
+// Draggable SVG Elements from
+// https://www.petercollingridge.co.uk/tutorials/svg/interactive/dragging/
+function makeDraggableSvg(svg) {
+    var selectedElement = false, offset, transform;
+    var diagram = new Diagram(svg);
+
+    svg.addEventListener('mousedown', startDrag);
+    svg.addEventListener('mousemove', drag);
+    svg.addEventListener('mouseup', endDrag);
+    svg.addEventListener('mouseleave', endDrag);
```

Comment:
what should we do there? it would make more sense to put it back to its original position, as it's difficult to drag it once it's on the edge. And/or should we put that event listener on the document? Then we would need to update the svg viewbox and/or size!

Reply:
I think the most intuitive action would be to end the dragging action just in boundary of the visible area, but we should draw the rectangle where the diagram is contained, to let the user know where it "ends"